cognite/v05/timeseries.py

```diff
@@ -146,68 +153,109 @@ def _get_datapoints_helper(name, aggregates=None, granularity=None, start=None,
+        latest_timestamp = int(datapoints[-1][-1]["timestamp"])
+        params["start"] = latest_timestamp + (_utils.granularity_to_ms(granularity) if granularity else 1)
     dps = []
     [dps.extend(el) for el in datapoints]
     return dps
 
 
-def _split_TimeseriesWithDatapoints_if_over_limit(timeseries_with_datapoints: TimeseriesWithDatapoints, limit: int) -> \
-        List[TimeseriesWithDatapoints]:
-    '''Takes a TimeseriesWithDatapoints and splits it into multiple so that each has a max number of datapoints equal
+def _get_datapoints_user_defined_limit(name, aggregates, granularity, start, end, limit, **kwargs):
+    """Returns a DatapointsResponse object with the requested data.
+
+    No paging or parallelizing is done.
+
+    Args:
+        name (str):       The name of the timeseries to retrieve data for.
+
+        aggregates (list):      The list of aggregate functions you wish to apply to the data. Valid aggregate functions
+                                are: 'average/avg, max, min, count, sum, interpolation/int, stepinterpolation/step'.
+
+        granularity (str):      The granularity of the aggregate values. Valid entries are : 'day/d, hour/h, minute/m,
+                                second/s', or a multiple of these indicated by a number as a prefix e.g. '12hour'.
+
+        start (Union[str, int, datetime]):    Get datapoints after this time. Format is N[timeunit]-ago where timeunit is w,d,h,m,s.
+                                    E.g. '2d-ago' will get everything that is up to 2 days old. Can also send time in ms since
+                                    epoch or a datetime object which will be converted to ms since epoch UTC.
+
+        end (Union[str, int, datetime]):      Get datapoints up to this time. Same format as for start.
+
+        limit (int):            Max number of dps to retrieve.
+
+    Keyword Arguments:
+        protobuf (bool):        Download the data using the binary protobuf format. Only applicable when getting raw data.
+                                Defaults to True.
+
+        api_key (str):          Your api-key.
+
+        project (str):          Project name.
+    Returns:
+        v05.dto.DatapointsResponse: A data object containing the requested data with several getter methods with different
+        output formats.
+    """
+    api_key, project = config.get_config_variables(kwargs.get("api_key"), kwargs.get("project"))
+    url = config.get_base_url(api_version=0.5) + "/projects/{}/timeseries/data/{}".format(project, quote(name, safe=""))
+
+    use_protobuf = kwargs.get("protobuf", True) and aggregates is None
+
+    params = {"aggregates": aggregates, "granularity": granularity, "limit": limit, "start": start, "end": end}
+
+    headers = {"api-key": api_key, "accept": "application/protobuf" if use_protobuf else "application/json"}
+    res = _utils.get_request(url, params=params, headers=headers)
```

Comment:
So, we use our API functionality to limit the number of datapoints? If this is the case, max limit is 100k. Then we should explicitly say that in the docstring of this method or caller method.

Reply:
And is this handled in this method when it’s not called with a limit? What happens if one of the intervals contains more that this threshold (100k)?

Reply:
@tuanng-cognite ok, will update the docstrings.
@henrikCognite this method is only used if the user passes limit to get_datapoints(). If not, it splits the request into equal time intervals and parallellizes the execution of _get_datapoints_helper() method which pages automatically within the each interval.